std/assembly/array.ts

```diff
@@ -476,6 +476,62 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    assert(isArray<T>());
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 1 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
+    result.dataLength = byteLength;
+    result.dataStart = dataStart;
+    // this statement retains the ArrayBuffer
+    result.data = changetype<ArrayBuffer>(dataStart);
+
+    // set the elements
+    let resultIndex: i32 = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      let childDataLength = child.dataLength;
```

Comment:
Hmm, you try access to `child.dataLength` before [null check](https://github.com/AssemblyScript/assemblyscript/pull/705/files#diff-3d69b4bb72041060ee7b8b49b0f22720R511)

Reply:
Whoops!